crates/chain-config/src/config/codec/encoder.rs

```diff
@@ -0,0 +1,304 @@
+                contract_state: create_encoder(path, "contract_state", compression)?,
+                contract_balance: create_encoder(path, "contract_balance", compression)?,
+            },
+        })
+    }
+
+    pub fn write_coins(&mut self, elements: Vec<CoinConfig>) -> anyhow::Result<()> {
+        match &mut self.encoder {
+            EncoderType::Json { buffer: state, .. } => {
+                state.coins.extend(elements);
+                Ok(())
+            }
+            EncoderType::Parquet { coins, .. } => coins.write(elements),
+        }
+    }
+
+    pub fn write_contracts(
+        &mut self,
+        elements: Vec<ContractConfig>,
+    ) -> anyhow::Result<()> {
+        match &mut self.encoder {
+            EncoderType::Json { buffer: state, .. } => {
+                state.contracts.extend(elements);
+                Ok(())
+            }
+            EncoderType::Parquet { contracts, .. } => contracts.write(elements),
+        }
+    }
+
+    pub fn write_messages(&mut self, elements: Vec<MessageConfig>) -> anyhow::Result<()> {
+        match &mut self.encoder {
+            EncoderType::Json { buffer: state, .. } => {
+                state.messages.extend(elements);
+                Ok(())
+            }
+            EncoderType::Parquet { messages, .. } => messages.write(elements),
+        }
+    }
+
+    pub fn write_contract_state(
+        &mut self,
+        elements: Vec<ContractStateConfig>,
+    ) -> anyhow::Result<()> {
+        match &mut self.encoder {
+            EncoderType::Json { buffer: state, .. } => {
+                state.contract_state.extend(elements);
+                Ok(())
+            }
+            EncoderType::Parquet { contract_state, .. } => contract_state.write(elements),
+        }
+    }
+
+    pub fn write_contract_balance(
+        &mut self,
+        elements: Vec<ContractBalance>,
+    ) -> anyhow::Result<()> {
+        match &mut self.encoder {
+            EncoderType::Json { buffer: state, .. } => {
+                state.contract_balance.extend(elements);
+                Ok(())
+            }
+            EncoderType::Parquet {
+                contract_balance, ..
+            } => contract_balance.write(elements),
+        }
+    }
+
+    pub fn close(self) -> anyhow::Result<()> {
+        match self.encoder {
+            EncoderType::Json {
+                buffer,
+                state_file_path,
+            } => {
+                let file = std::fs::File::create(state_file_path)?;
+                serde_json::to_writer(file, &buffer)?;
+                Ok(())
+            }
+            EncoderType::Parquet {
+                coins,
+                messages,
+                contracts,
+                contract_state,
+                contract_balance,
+            } => {
+                coins.close()?;
+                messages.close()?;
+                contracts.close()?;
+                contract_state.close()?;
+                contract_balance.close()?;
```

Comment:
What kind of error we can get here? If closing `coins` fails, how does it affect other encoders?

Reply:
If `coins` fails nothing else will get closed. 

Closing a parquet file is to flush any buffers and to write the metadata at the end of the file.

Seeing as a snapshot is only complete if all of the parts are properly written, didn't see a need to even try to close other files if one fails.

I'll probably go ahead and parallelize this file writing rn. If I succeed then I will have to handle each failure separately and not short-circuit them.